File: tools/fix_heist_icon_alpha.py
#!/usr/bin/env python3
"""Remove flat backgrounds from heist/hacking inventory PNGs and resize to 256px."""
from __future__ import annotations

import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("pip install Pillow", file=sys.stderr)
    sys.exit(1)

IMAGES_DIR = Path(__file__).resolve().parents[1] / "resources" / "[qb]" / "qb-inventory" / "html" / "images"
TARGET = 256

NAMES = (
    "basic_tablet",
    "advanced_tablet",
    "military_tablet",
    "basic_flashdrive",
    "encrypted_flashdrive",
    "military_flashdrive",
    "tow_chain",
    "drill",
    "thermite",
    "security_card_01",
    "security_card_02",
)


def corner_bg_color(img: Image.Image) -> tuple[int, int, int]:
    w, h = img.size
    samples: list[tuple[int, int, int]] = []
    for x, y in ((1, 1), (w - 2, 1), (1, h - 2), (w - 2, h - 2), (w // 2, 1), (w // 2, h - 2)):
        r, g, b, _ = img.getpixel((x, y))
        samples.append((r, g, b))
    return (
        sum(s[0] for s in samples) // len(samples),
        sum(s[1] for s in samples) // len(samples),
        sum(s[2] for s in samples) // len(samples),
    )


def color_dist(a: tuple[int, int, int], b: tuple[int, int, int]) -> float:
    return ((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2) ** 0.5


def process(path: Path, tolerance: float = 55.0) -> bool:
    img = Image.open(path).convert("RGBA")
    if max(img.size) != TARGET:
        img.thumbnail((TARGET, TARGET), Image.Resampling.LANCZOS)
        canvas = Image.new("RGBA", (TARGET, TARGET), (0, 0, 0, 0))
        ox = (TARGET - img.width) // 2
        oy = (TARGET - img.height) // 2
        canvas.paste(img, (ox, oy), img)
        img = canvas
    bg = corner_bg_color(img)
    w, h = img.size
    px = img.load()
    changed = False
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            if a == 0:
                continue
            rgb = (r, g, b)
            if color_dist(rgb, bg) <= tolerance:
                px[x, y] = (r, g, b, 0)
                changed = True
                continue
            if r > 160 and g > 160 and b > 160 and max(r, g, b) - min(r, g, b) < 30:
                px[x, y] = (r, g, b, 0)
                changed = True
            elif r > 200 and g > 200 and b > 200:
                px[x, y] = (r, g, b, 0)
                changed = True
    img.save(path, optimize=True)
    return changed


def main() -> int:
    if not IMAGES_DIR.is_dir():
        print(f"Missing: {IMAGES_DIR}", file=sys.stderr)
        return 1
    touched = 0
    for name in NAMES:
        path = IMAGES_DIR / f"{name}.png"
        if not path.is_file():
            print(f"missing {path.name}")
            continue
        if process(path):
            print(f"fixed {path.name}")
            touched += 1
        else:
            print(f"processed {path.name}")
            touched += 1
    print(f"Done. Updated {touched} file(s).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/fix_heist_icon_alpha.mjs
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { PNG } from 'pngjs';

const __dir = dirname(fileURLToPath(import.meta.url));
const IMAGES = join(__dir, '..', 'resources', '[qb]', 'qb-inventory', 'html', 'images');
const TARGET = 256;
const NAMES = [
  'basic_tablet', 'advanced_tablet', 'military_tablet',
  'basic_flashdrive', 'encrypted_flashdrive', 'military_flashdrive',
  'tow_chain', 'drill', 'thermite', 'security_card_01', 'security_card_02',
];

function dist(a, b) {
  return Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);
}

function cornerBg(data, w, h) {
  const pts = [[1, 1], [w - 2, 1], [1, h - 2], [w - 2, h - 2], [Math.floor(w / 2), 1], [Math.floor(w / 2), h - 2]];
  let r = 0, g = 0, b = 0;
  for (const [x, y] of pts) {
    const i = (w * y + x) << 2;
    r += data[i]; g += data[i + 1]; b += data[i + 2];
  }
  return [Math.round(r / pts.length), Math.round(g / pts.length), Math.round(b / pts.length)];
}

function processPng(buf) {
  const src = PNG.sync.read(buf);
  let { width: w, height: h, data } = src;
  const scale = TARGET / Math.max(w, h);
  if (scale !== 1) {
    const nw = Math.max(1, Math.round(w * scale));
    const nh = Math.max(1, Math.round(h * scale));
    const tmp = new PNG({ width: nw, height: nh });
    for (let y = 0; y < nh; y++) {
      for (let x = 0; x < nw; x++) {
        const sx = Math.min(w - 1, Math.floor(x / scale));
        const sy = Math.min(h - 1, Math.floor(y / scale));
        const si = (w * sy + sx) << 2;
        const di = (nw * y + x) << 2;
        tmp.data[di] = data[si];
        tmp.data[di + 1] = data[si + 1];
        tmp.data[di + 2] = data[si + 2];
        tmp.data[di + 3] = data[si + 3];
      }
    }
    w = nw; h = nh; data = tmp.data;
  }
  const out = new PNG({ width: TARGET, height: TARGET });
  out.data.fill(0);
  const ox = Math.floor((TARGET - w) / 2);
  const oy = Math.floor((TARGET - h) / 2);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const si = (w * y + x) << 2;
      const di = (TARGET * (oy + y) + (ox + x)) << 2;
      out.data[di] = data[si];
      out.data[di + 1] = data[si + 1];
      out.data[di + 2] = data[si + 2];
      out.data[di + 3] = data[si + 3];
    }
  }
  const bg = cornerBg(out.data, TARGET, TARGET);
  const tol = 55;
  for (let i = 0; i < out.data.length; i += 4) {
    const r = out.data[i], g = out.data[i + 1], b = out.data[i + 2];
    if (out.data[i + 3] === 0) continue;
    const rgb = [r, g, b];
    if (dist(rgb, bg) <= tol) { out.data[i + 3] = 0; continue; }
    if (r > 160 && g > 160 && b > 160 && Math.max(r, g, b) - Math.min(r, g, b) < 30) out.data[i + 3] = 0;
    else if (r > 200 && g > 200 && b > 200) out.data[i + 3] = 0;
  }
  return PNG.sync.write(out);
}

let n = 0;
for (const name of NAMES) {
  const p = join(IMAGES, `${name}.png`);
  if (!existsSync(p)) { console.log('missing', name); continue; }
  writeFileSync(p, processPng(readFileSync(p)));
  console.log('fixed', name);
  n++;
}
console.log(`Done. ${n} file(s).`);
